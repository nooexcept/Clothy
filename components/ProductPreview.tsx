import React from 'react'
import { makeStyles, createStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import Zoom from '@material-ui/core/Zoom'
import CardActionArea from '@material-ui/core/CardActionArea'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Typography from '@material-ui/core/Typography'

const useStyles = makeStyles(() =>
    createStyles({
        root: {
            height: '17vh',
            minHeight: '100px',
            overflow: 'hidden',
        },
        cardRoot: {
            display: 'flex',
            height: '100%',
        },
        details: {
            display: 'flex',
            flexDirection: 'column',
        },
        content: {
            flex: '1 0 auto',
        },
    })
)

interface ProductPreviewProps {
    name: string
    desc: string
    url: string
    img: string
}

const ProductPreview: React.SFC<ProductPreviewProps> = ({
    name,
    desc,
    url,
    img,
}: ProductPreviewProps) => {
    const classes = useStyles()

    return (
        <Zoom in>
            <Paper
                onClick={() => (window.location.href = `/products/${url}`)}
                className={classes.root}
                elevation={3}
            >
                <CardActionArea className={classes.cardRoot}>
                    <div className={classes.details}>
                        <CardContent className={classes.content}>
                            <Typography component="h6" variant="h6">
                                {name}
                            </Typography>
                            <Typography
                                variant="subtitle1"
                                color="textSecondary"
                            >
                                {desc}
                            </Typography>
                        </CardContent>
                    </div>
                    <CardMedia
                        style={{ width: 250, height: '100%' }}
                        image={img}
                        title={name}
                    />
                </CardActionArea>
            </Paper>
        </Zoom>
    )
}

export default ProductPreview
