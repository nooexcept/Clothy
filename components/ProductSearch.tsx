import React, { useState } from 'react'
import axios, { CancelTokenSource } from 'axios'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import { fade, makeStyles, Theme, createStyles } from '@material-ui/core/styles'
import SearchIcon from '@material-ui/icons/Search'
import Box from '@material-ui/core/Box'
import Grid from '@material-ui/core/Grid'
import InputBase from '@material-ui/core/InputBase'
import ProductPreview from './ProductPreview'
import { withLimitedWords } from '../src/util'
import { Product } from '../src/types'

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            position: 'relative',
        },
        search: {
            position: 'relative',
            borderRadius: theme.shape.borderRadius,
            backgroundColor: fade(theme.palette.common.black, 0.1),
            '&:hover': {
                backgroundColor: fade(theme.palette.common.black, 0.15),
            },
            marginRight: theme.spacing(2),
            marginLeft: 0,
            width: '100%',
            [theme.breakpoints.up('sm')]: {
                marginLeft: theme.spacing(3),
                width: 'auto',
            },
        },
        searchIcon: {
            padding: theme.spacing(0, 2),
            height: '100%',
            position: 'absolute',
            pointerEvents: 'none',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
        },
        inputRoot: {
            color: 'inherit',
        },
        inputInput: {
            padding: theme.spacing(1, 1, 1, 0),
            paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
            transition: theme.transitions.create('width'),
            width: '100%',
            [theme.breakpoints.up('md')]: {
                width: '20ch',
            },
        },
        suggestions: {
            position: 'absolute',
            width: '80%',
            marginLeft: '24px',
            marginRight: '16px',
            marginTop: '8px',
        },
    })
)

const ProductSearch: React.FC = () => {
    const classes = useStyles()

    const [searchText, setSearchText] = useState('')
    const [suggestions, setSuggestions] = useState({ display: false, data: [] })

    const CancelToken = axios.CancelToken

    let source: CancelTokenSource | null = null

    const OnSearch = async (e: React.ChangeEvent<HTMLInputElement>) => {
        const textVal = e.target.value
        setSearchText(textVal)

        if (!!source) {
            source.cancel()
        }

        if (textVal && textVal.length > 2) {
            source = CancelToken.source()

            try {
                const suggested = await axios.get(
                    `${process.env.NEXT_PUBLIC_API_URL}/api/products/search/${textVal}/0/3`,
                    {
                        cancelToken: source.token,
                    }
                )

                setSuggestions({
                    display: true,
                    data: suggested.data.map((prod: Product) => ({
                        ...prod,
                        description: `${withLimitedWords(
                            prod.description,
                            12
                        )}...`,
                    })),
                })
            } catch (e) {
                source = null
                setSuggestions({ display: false, data: [] })
            }
        } else {
            source = null
            setSuggestions({ display: false, data: [] })
        }
    }

    return (
        <ClickAwayListener
            onClickAway={() =>
                setSuggestions({ ...suggestions, display: false })
            }
        >
            <Box className={classes.root}>
                <div className={classes.search}>
                    <div className={classes.searchIcon}>
                        <SearchIcon />
                    </div>
                    <InputBase
                        placeholder="Search..."
                        classes={{
                            root: classes.inputRoot,
                            input: classes.inputInput,
                        }}
                        inputProps={{ 'aria-label': 'search' }}
                        value={searchText}
                        onChange={OnSearch}
                        onClick={() =>
                            setSuggestions({ ...suggestions, display: true })
                        }
                    />
                </div>
                {suggestions.display && suggestions.data.length > 0 && (
                    <div id="suggestions" className={classes.suggestions}>
                        <Grid
                            container
                            direction="column"
                            justify="center"
                            alignItems="center"
                            spacing={1}
                        >
                            {suggestions.data.map((prod: Product) => (
                                <Grid item key={prod.url}>
                                    <ProductPreview
                                        name={prod.name}
                                        desc={prod.description}
                                        url={prod.url}
                                        img={prod.images[0]}
                                    />
                                </Grid>
                            ))}
                        </Grid>
                    </div>
                )}
            </Box>
        </ClickAwayListener>
    )
}

export default ProductSearch
