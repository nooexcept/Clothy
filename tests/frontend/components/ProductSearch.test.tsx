import React from 'react'
import { ReactWrapper, HTMLAttributes } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { createMount } from '@material-ui/core/test-utils'
import ProductPreview from '../../../components/ProductPreview'
import ProductSearch from '../../../components/ProductSearch'
import InputBase from '@material-ui/core/InputBase'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'

const mockedSuggestions = [
    {
        name: 'Ultra chill jumpers',
        url: 'ultra-chill-jumpers',
        description: 'If you wanna look chill, you need ultra chill jumpers',
        stock: 6,
        price: 80,
        cod: 'JUMPERS002',
        images: ['/photo-of-woman-standing-on-staircase-3120339.jpg'],
    },
]

const altMockedSuggestions = [
    {
        name: 'This is wrong',
        url: 'should-not-show',
        description: 'This is only displayed if some test failed',
        stock: 0,
        price: 0,
        cod: 'FAIL001',
        images: ['/failure.jpg'],
    },
]

const searchEndpoint = (txt: string) =>
    `${process.env.NEXT_PUBLIC_API_URL}/api/products/search/${txt}/0/3`

let wrapper: ReactWrapper
let mount: ReturnType<typeof createMount>
const mock = new MockAdapter(axios)

beforeEach(() => {
    mount = createMount()
    wrapper = mount(<ProductSearch />)
})

afterEach(() => {
    mock.reset()
    wrapper.unmount()
})

describe('<ProductSearch /> rendering', () => {
    it('should render without problems', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it("shouldn't show suggestions if there is no text", () => {
        const suggestionsDiv = wrapper.find('#suggestions')
        expect(suggestionsDiv.length).toEqual(0)
    })
})

describe('<ProductSearch /> interactions', () => {
    it('should show suggestions only when appropriate', async () => {
        mock.onGet(searchEndpoint('Ultra')).reply(200, mockedSuggestions)

        await act(async () => {
            wrapper
                .find(InputBase)
                .at(0)
                .find('input')
                .simulate('change', { target: { value: 'Ultra' } })
        })

        wrapper.update()

        expect(wrapper.find('#suggestions').length).toEqual(1)

        await act(async () => {
            ;(wrapper.find('ClickAwayListener').props() as HTMLAttributes & {
                onClickAway: () => void
            }).onClickAway()
        })

        wrapper.update()

        expect(wrapper.find('#suggestions').length).toEqual(0)

        await act(async () => {
            wrapper.find(InputBase).at(0).find('input').simulate('click')
        })

        wrapper.update()

        expect(wrapper.find('#suggestions').length).toEqual(1)
    })

    it('should show the latest data requested', async () => {
        mock.onGet(searchEndpoint('This')).reply(200, mockedSuggestions)

        mock.onGet(searchEndpoint('That')).reply(() => {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve([200, altMockedSuggestions])
                }, 1000)
            })
        })

        const textInput = wrapper.find(InputBase).at(0).find('input')

        await act(async () => {
            textInput.simulate('change', { target: { value: 'That' } })
            textInput.simulate('change', { target: { value: 'This' } })
        })

        wrapper.update()

        const suggestionsDiv = wrapper.find('#suggestions')
        expect(suggestionsDiv.length).toEqual(1)

        const correctPreviewProps = wrapper.find(ProductPreview).props()

        expect(correctPreviewProps).toEqual({
            name: mockedSuggestions[0].name,
            desc: `${mockedSuggestions[0].description}...`,
            url: mockedSuggestions[0].url,
            img: mockedSuggestions[0].images[0],
        })
    })

    it('shouldn\'t show suggestions if there is not enough text', async () => {
        mock.onGet(searchEndpoint('')).reply(200, [])
        const textInput = wrapper.find(InputBase).at(0).find('input')

        await act(async () => {
            textInput.simulate('change', { target: { value: '' } })
        })

        wrapper.update()
        const suggestionsDiv = wrapper.find('#suggestions')
        expect(suggestionsDiv.length).toEqual(0)
    })
})
