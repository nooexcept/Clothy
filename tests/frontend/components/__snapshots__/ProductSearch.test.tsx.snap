// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSearch /> rendering should render without problems 1`] = `
<ProductSearch>
  <ClickAwayListener
    onClickAway={[Function]}
  >
    <Styled(MuiBox)
      className="makeStyles-root-1"
      onClick={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="MuiBox-root MuiBox-root-7 makeStyles-root-1"
        onClick={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className="makeStyles-search-2"
        >
          <div
            className="makeStyles-searchIcon-3"
          >
            <ForwardRef>
              <WithStyles(ForwardRef(SvgIcon))>
                <ForwardRef(SvgIcon)
                  classes={
                    Object {
                      "colorAction": "MuiSvgIcon-colorAction",
                      "colorDisabled": "MuiSvgIcon-colorDisabled",
                      "colorError": "MuiSvgIcon-colorError",
                      "colorPrimary": "MuiSvgIcon-colorPrimary",
                      "colorSecondary": "MuiSvgIcon-colorSecondary",
                      "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                      "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                      "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                      "root": "MuiSvgIcon-root",
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                  </svg>
                </ForwardRef(SvgIcon)>
              </WithStyles(ForwardRef(SvgIcon))>
            </ForwardRef>
          </div>
          <WithStyles(ForwardRef(InputBase))
            classes={
              Object {
                "input": "makeStyles-inputInput-5",
                "root": "makeStyles-inputRoot-4",
              }
            }
            inputProps={
              Object {
                "aria-label": "search",
              }
            }
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Search..."
            value=""
          >
            <ForwardRef(InputBase)
              classes={
                Object {
                  "adornedEnd": "MuiInputBase-adornedEnd",
                  "adornedStart": "MuiInputBase-adornedStart",
                  "colorSecondary": "MuiInputBase-colorSecondary",
                  "disabled": "Mui-disabled",
                  "error": "Mui-error",
                  "focused": "Mui-focused",
                  "formControl": "MuiInputBase-formControl",
                  "fullWidth": "MuiInputBase-fullWidth",
                  "input": "MuiInputBase-input makeStyles-inputInput-5",
                  "inputAdornedEnd": "MuiInputBase-inputAdornedEnd",
                  "inputAdornedStart": "MuiInputBase-inputAdornedStart",
                  "inputHiddenLabel": "MuiInputBase-inputHiddenLabel",
                  "inputMarginDense": "MuiInputBase-inputMarginDense",
                  "inputMultiline": "MuiInputBase-inputMultiline",
                  "inputTypeSearch": "MuiInputBase-inputTypeSearch",
                  "marginDense": "MuiInputBase-marginDense",
                  "multiline": "MuiInputBase-multiline",
                  "root": "MuiInputBase-root makeStyles-inputRoot-4",
                }
              }
              inputProps={
                Object {
                  "aria-label": "search",
                }
              }
              onChange={[Function]}
              onClick={[Function]}
              placeholder="Search..."
              value=""
            >
              <div
                className="MuiInputBase-root makeStyles-inputRoot-4"
                onClick={[Function]}
              >
                <input
                  aria-label="search"
                  className="MuiInputBase-input makeStyles-inputInput-5"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Search..."
                  type="text"
                  value=""
                />
              </div>
            </ForwardRef(InputBase)>
          </WithStyles(ForwardRef(InputBase))>
        </div>
      </div>
    </Styled(MuiBox)>
  </ClickAwayListener>
</ProductSearch>
`;
